File: Examples/Comparison/prune_reroot_compare.py
from copy import deepcopy
from time import perf_counter

from Bio import Phylo
from Bio.Align import MultipleSeqAlignment
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor


def build_records(sequence_count=80, width=140):
    template = ("ACGT" * ((width + 3) // 4))[:width]
    bases = "ACGT"
    records = []
    for sequence_index in range(1, sequence_count + 1):
        chars = list(template)
        for position in range(1, width + 1):
            if (sequence_index + position) % 11 == 0:
                chars[position - 1] = bases[(sequence_index + position - 2) % len(bases)]
            elif (sequence_index * position) % 17 == 0:
                chars[position - 1] = bases[(sequence_index + 2 * position - 3) % len(bases)]
        records.append(SeqRecord(Seq("".join(chars)), id=f"seq{sequence_index}", description="prune_input"))
    return records


def time_call(func):
    start = perf_counter()
    result = func()
    return (perf_counter() - start) * 1000.0, result


def build_tree():
    alignment = MultipleSeqAlignment(build_records())
    calculator = DistanceCalculator("identity")
    constructor = DistanceTreeConstructor(calculator)
    distance_matrix = calculator.get_distance(alignment)
    return constructor.nj(distance_matrix)


def prune_tree(tree, keep_names):
    rooted = deepcopy(tree)
    keep = set(keep_names)
    for terminal in list(rooted.get_terminals()):
        if terminal.name not in keep:
            rooted.prune(terminal.name)
    return rooted


def reroot_tree(tree):
    rooted = deepcopy(tree)
    rooted.root_with_outgroup("seq1")
    return rooted


def main():
    tree = build_tree()
    keep_names = [f"seq{index}" for index in range(1, 41, 2)]

    prune_ms, pruned = time_call(lambda: prune_tree(tree, keep_names))
    reroot_ms, rerooted = time_call(lambda: reroot_tree(tree))

    assert len(pruned.get_terminals()) == len(keep_names)
    assert any(terminal.name == "seq1" for terminal in rerooted.get_terminals())

    print(f"python_prune_ms={prune_ms:.4f}")
    print(f"python_reroot_ms={reroot_ms:.4f}")
    print(f"python_prune_ok={str(True).lower()}")
    print(f"python_reroot_ok={str(True).lower()}")


if __name__ == "__main__":
    main()

File: Examples/Comparison/prune_reroot_compare.jl
pushfirst!(LOAD_PATH, abspath(joinpath(@__DIR__, "..", "..")))

using Random
using BioToolkit
include(joinpath(@__DIR__, "..", "..", "scripts", "benchmark_helpers.jl"))

function build_tree(sequence_count::Integer=80, width::Integer=140)
    Random.seed!(31415)
    template = repeat("ACGT", cld(width, 4))[1:width]
    bases = ('A', 'C', 'G', 'T')
    sequences = String[]

    for sequence_index in 1:sequence_count
        chars = collect(template)
        for position in 1:width
            if mod(sequence_index + position, 11) == 0
                chars[position] = bases[mod(sequence_index + position - 2, length(bases)) + 1]
            elseif mod(sequence_index * position, 17) == 0
                chars[position] = bases[mod(sequence_index + 2 * position - 3, length(bases)) + 1]
            end
        end
        push!(sequences, String(chars))
    end

    labels = ["seq$(index)" for index in eachindex(sequences)]
    distances = BioToolkit.distance_matrix(sequences; method=:hamming)
    return BioToolkit.neighbor_joining_tree(distances, labels)
end

function main()
    tree = build_tree()
    keep_names = ["seq$(index)" for index in 1:2:40]

    prune_ms, pruned = repeat_elapsed_ms(() -> BioToolkit.prune(tree, keep_names), 100)
    reroot_ms, rerooted = repeat_elapsed_ms(() -> BioToolkit.root_with_outgroup(tree, "seq1"), 100)

    @assert length(BioToolkit.get_terminals(pruned)) == length(keep_names)
    @assert occursin("seq1", BioToolkit.write_newick(rerooted))

    python_output = read(`conda run -n general --no-capture-output python $(joinpath(@__DIR__, "prune_reroot_compare.py"))`, String)
    python_prune_ms = parse(Float64, match(r"python_prune_ms=([0-9.]+)", python_output).captures[1])
    python_reroot_ms = parse(Float64, match(r"python_reroot_ms=([0-9.]+)", python_output).captures[1])
    python_prune_ok = match(r"python_prune_ok=(true|false)", python_output).captures[1] == "true"
    python_reroot_ok = match(r"python_reroot_ok=(true|false)", python_output).captures[1] == "true"

    @assert python_prune_ok
    @assert python_reroot_ok

    println("Pruning and rerooting benchmark")
    println("  sequence_count=80")
    println("  julia_prune_ms=$(round(prune_ms / 100; digits=4))")
    println("  julia_reroot_ms=$(round(reroot_ms / 100; digits=4))")
    println("  python_prune_ms=$(round(python_prune_ms; digits=4))")
    println("  python_reroot_ms=$(round(python_reroot_ms; digits=4))")
    print(python_output)
end

main()